client/state/data-layer/wpcom/videos/poster/index.js

```diff
@@ -0,0 +1,51 @@
+/**
+ * Internal dependencies
+ */
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import { VIDEO_EDITOR_POSTER_UPDATE } from 'state/action-types';
+import {
+	updateVideoEditorPosterSuccess,
+	updateVideoEditorPosterFailure,
+	updatingVideoEditorPoster,
+} from 'state/ui/editor/video-editor/actions';
+
+/**
+ * Updates the poster for a video.
+ *
+ * @param  {Object} store  Redux store
+ * @param  {Object} action  Action object
+ * @param {Function} next  Dispatches to next middleware in chain
+ * @returns {Object} original action
+ */
+export const updateVideoEditorPoster = ( { dispatch }, action, next ) => {
+	if ( ! ( 'file' in action.params ) && ! ( 'at_time' in action.params ) ) {
+		return next( action );
+	}
+
+	dispatch( updatingVideoEditorPoster() );
+
+	dispatch( http( {
+		apiVersion: '1.1',
+		method: 'POST',
+		path: `/videos/${ action.videoId }/poster`,
+		body: 'at_time' in action.params ? action.params : undefined,
+		formData: 'file' in action.params ? [ [ 'poster', action.params.file ] ] : undefined,
+		onSuccess: action,
+		onFailure: action,
+	} ) );
+
+	return next( action );
+};
+
+export const updateVideoPosterSuccess = ( { dispatch }, action, next, { poster } ) => {
+	dispatch( updateVideoEditorPosterSuccess( poster ) );
+};
+
+export const updateVideoPosterError = ( { dispatch } ) => {
+	dispatch( updateVideoEditorPosterFailure() );
+};
+
+export default {
+	[ VIDEO_EDITOR_POSTER_UPDATE ]: [ dispatchRequest( updateVideoEditorPoster, updateVideoPosterSuccess, updateVideoPosterError ) ],
```

Comment:
@donnapep I'm not sure what your question is. the data layer latches into existing actions which feed through the system. it handle side-effects and data fetching.

the idea is that Calypso can function as best as possible even if no data gets synchronized. so maybe we fire an event `VIDEO_POSTER_UPDATE` which specifies a video and a frame for the post (is that what this is about?) then Calypso can hold onto that state the same way it would if it had just loaded that data from the API.

in this case the data layer sees that action fly by and it says to itself, "self, I know that when someone changes the video poster we also need to let WordPress.com know so that it persists and propagates, I will do just that"

now, once we have fully uploaded the data and get some response we can make a judgement. if it's as simple as saying, "use frame 1337 for the poster" then if we get a positive response back from the server showing that it stored 1337 and our state already shows 1337 (because Redux state operates independently of the middleware) then we don't actually need to do a thing. no changes are required. on the other hand, if it comes back with the wrong frame number we may need to respond in one of many possible ways: retry the operation (Calypso meanwhile is unaware of the failure), report the failure to the user, or just force Calypso to update (overwriting the customer's choice).

please let me know if this is addressing your question!